File: NeetCode 150/Binary Tree/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_sum = float('-inf')

        def dfs(node):
            if not node:
                return 0
            
            left_gain = max(dfs(node.left), 0)
            right_gain = max(dfs(node.right),0)

            #Two Possibilities:
            #1. If this node is part of the path which includes left and right childs, for which we update max_sum
            path_sum = node.val + left_gain + right_gain
            self.max_sum = max(self.max_sum,path_sum)

            #2. Send Gain in curr path to higher node, where current path goes from left child or right which ever is more.
            return node.val + max(left_gain,right_gain)
        
        dfs(root)
        return self.max_sum #Valid for case 1 and case 2 as in case 2 this may or maynot be included but max_sum is already computed

File: NeedCode 150 (JAVA)/Trees/124. Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int max_sum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        dfs(root);
        return max_sum;
    }
    private int dfs(TreeNode node){
        if(node == null) return 0;

        int left_gain = Math.max(dfs(node.left),0);
        int right_gain = Math.max(dfs(node.right),0);

        int path_sum = node.val+left_gain+right_gain;
        max_sum = Math.max(max_sum,path_sum);

        return node.val + Math.max(left_gain,right_gain);
    }
}
